### PR TITLE
Add pattern to do affine.min canonicalization after tile and distribute

### PR DESCRIPTION
**Pull request by ThomasRaoux**

### CODE REVIEW

File: iree/compiler/Conversion/Common/Transforms.h [hunk 1]
```diff
@@ -103,6 +104,24 @@ LogicalResult tileAndFuseLinalgBufferOps(
...
+/// distribute.
+/// Try to fold a affine.min op by matching the following form:
+/// ```
+/// scf.for %iv = %lb to %ub step %step
+///   %affine.min affine_map<(d0, d1) -> (N, d0 - d1)>(%ub, %iv)
+/// ```
+/// With N a compile time constant. This operations can be replace by
+/// `%cN = constant N : index` if we can prove that %lb, %step and %ub are
+/// divisible by N.
+struct AffineMinDistributedSCFCanonicalizationPattern
```
Comment (MaheshRavishankar) [inline iree/compiler/Conversion/Common/Transforms.h:117]: I'd rather not expose the struct directly, but rather use an "populateAffineMinSCFCanonicalizationPattern" method.

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]

File: iree/compiler/Conversion/Common/Transforms.h (hunk 1 above)
Comment (ThomasRaoux) [reply iree/compiler/Conversion/Common/Transforms.h]: Make sense, I moved the structure into the cpp file

File: iree/compiler/Conversion/Common/Transforms.cpp [hunk 2]
```diff
@@ -443,5 +443,164 @@ LogicalResult tileAndFuseLinalgBufferOps(
...
+
+/// Return true if we can prove that affineMinOp result is positive and
+/// divisible by the given |dividend|. This is true if all the the results of
+/// the associated affine map are positive and divisible by |dividend|.
+/// This speciically look for the following pattern:
+/// ```
+/// scf.for %iv = %lb to %ub step %step
+///   %affine.min affine_map<(d0, d1) -> (N, d0 - d1)>(%ub, %iv)
+/// ```
+static bool affineMinOpDivisible(AffineMinOp minOp, int64_t dividend) {
```
Comment (MaheshRavishankar) [inline iree/compiler/Conversion/Common/Transforms.cpp:456]: Nit about name: This sounds like the affine_min is divisible by tile size :) . If you cant think of anything better, I cant either.

**Review by MaheshRavishankar — APPROVED** [review]

File: iree/compiler/Conversion/Common/Transforms.cpp (hunk 2 above)
Comment (ThomasRaoux) [reply iree/compiler/Conversion/Common/Transforms.cpp:456]: Well it does mean the result of the affine_min is divisible by tile size. We check that both side of the min are divisible to prove that the result is divisible. If it wasn't we couldn't do the transformation. Maybe I misunderstood your comment.

Comment (MaheshRavishankar) [reply iree/compiler/Conversion/Common/Transforms.cpp:456]: Its a minor nit. I already approved. My read was that "affine min is divisible" which to me is a strange name. I didnt parse it the same as "result of affine min is divisible".
Like I said, very minor issue. You can ignore it if this makes no sense to you.

**Pull request merged**